src/index.ts

```diff
@@ -1,10 +1,273 @@
-import * as plugin from './ts_plugin';
+import * as ts_module from "../node_modules/typescript/lib/tsserverlibrary";
+import * as tslint from 'tslint';
+import * as path from 'path';
 
-function factory(mod: { typescript: any /*typeof  ts*/ }) {
-  plugin.init(mod);
-  return plugin;
+interface Map<V> {
+    [key: string]: V;
+}
+
+let codeFixActions: Map<Map<tslint.Fix>> = Object.create(null);
+
+function computeKey(start: number, end: number): string {
+    return `[${start},${end}]`;
+}
+
+let configFile: string = null;
+let configuration: tslint.Configuration.IConfigurationFile = null;
+let isTsLint4: boolean = true;
+
+let configCache = {
+    filePath: <string>null,
+    configuration: <any>null,
+    isDefaultConfig: false,
+    configFilePath: <string>null
 };
 
-factory['create'] = plugin.create;
+let linter: typeof tslint.Linter = null;
+let linterConfiguration: typeof tslint.Configuration = null;
+
+let value = tslint;
+linter = value.Linter;
+linterConfiguration = value.Configuration;
+
+if (!isTsLint4) {
+    //linter = value;
+}
+/*function isTsLintVersion4(linter) {
+    let version = '1.0.0';
+    try {
+        version = linter.VERSION;
+    } catch (e) {
+    }
+    return semver.satisfies(version, ">= 4.0.0 || >= 4.0.0-dev");
+}*/
+
+function init(modules: { typescript: typeof ts_module }) {
+  const ts = modules.typescript;
+
+  function create(info: ts.server.PluginCreateInfo) {
+
+    info.project.projectService.logger.info("tslint-language-service loaded");
+
+    // Set up decorator
+    const proxy = Object.create(null) as ts.LanguageService;
+    const oldLS = info.languageService;
+    for (const k in oldLS) {
+      (<any>proxy)[k] = function () {
+        return oldLS[k].apply(oldLS, arguments);
+      }
+    }
+
+    function makeDiagnostic(problem: tslint.RuleFailure, file: ts.SourceFile): ts.Diagnostic {
+        let message = (problem.getRuleName() !== null)
+            ? `${problem.getFailure()} (${problem.getRuleName()})`
+            : `${problem.getFailure()}`;
+
+        // tslint5 supports to assign severities to rules
+        let category;
+        if ((<any>problem).getRuleSeverity && (<any>problem).getRuleSeverity() === 'error') { 
+            category = ts.DiagnosticCategory.Error;
+        } else {
+            category = ts.DiagnosticCategory.Warning;
+        }
+
+        let diagnostic: ts.Diagnostic = {
+            file: file,
+            start: problem.getStartPosition().getPosition(),
+            length: problem.getEndPosition().getPosition() - problem.getStartPosition().getPosition(),
+            messageText: message,
+            category: category,
+            //TODO we "steal"" an error code with a registered code fix. 2515 = implement inherited abstract class
+            code: 2515
```

Comment:
yes, this would be better. 

A downside is that then you have to duplicate the `codefix` namespace. However, this is an OK workaround for not having an API to register code fixes. Adding a comment why this code is duplicated would make me feel better ;-)